lib/pos_fee_estimator.go

```diff
@@ -194,19 +210,44 @@ func (posFeeEstimator *PoSFeeEstimator) removeBlockNoLock(block *MsgDeSoBlock) e
 func (posFeeEstimator *PoSFeeEstimator) UpdateGlobalParams(globalParams *GlobalParamsEntry) error {
 	posFeeEstimator.rwLock.Lock()
 	defer posFeeEstimator.rwLock.Unlock()
+
+	// Create a temporary transaction register to test the new global params.
 	tempTransactionRegister := NewTransactionRegister()
 	tempTransactionRegister.Init(globalParams.Copy())
+
 	for _, block := range posFeeEstimator.cachedBlocks {
 		if err := addBlockToTransactionRegister(tempTransactionRegister, block); err != nil {
 			return errors.Wrap(err, "PosFeeEstimator.UpdateGlobalParams: error adding block to tempTransactionRegister")
 		}
 	}
+
+	posFeeEstimator.globalParams = globalParams
 	return nil
 }
 
-// sortCachedBlocks sorts the cached blocks by height & tstamp just to be safe.
-func (posFeeEstimator *PoSFeeEstimator) sortCachedBlocks() {
-	posFeeEstimator.cachedBlocks = collections.SortStable(posFeeEstimator.cachedBlocks,
+// cleanUpPastBlocks cleans up the input blocks slice, deduping, sorting, and pruning the blocks by height.
+func (posFeeEstimator *PoSFeeEstimator) cleanUpPastBlocks(blocks []*MsgDeSoBlock) []*MsgDeSoBlock {
+	dedupedBlocks := posFeeEstimator.dedupeBlocksByBlockHeight(blocks)
+	sortedBlocks := posFeeEstimator.sortBlocksByBlockHeightAndTstamp(dedupedBlocks)
+	return posFeeEstimator.pruneBlocksToMaxNumPastBlocks(sortedBlocks)
+}
+
+// dedupeBlocksByBlockHeight deduplicates the blocks by block height. If multiple blocks have the same
+// height, it keeps the one with the highest view.
+func (posFeeEstimator *PoSFeeEstimator) dedupeBlocksByBlockHeight(blocks []*MsgDeSoBlock) []*MsgDeSoBlock {
+	blocksByBlockHeight := make(map[uint64]*MsgDeSoBlock)
+	for _, block := range blocks {
+		existingBlock, hasExitingBlock := blocksByBlockHeight[block.Header.Height]
+		if !hasExitingBlock || existingBlock.Header.GetView() < block.Header.GetView() {
+			blocksByBlockHeight[block.Header.Height] = block
+		}
+	}
+	return collections.MapValues(blocksByBlockHeight)
```

Comment:
Should cached blocks be a map instead of a slice? Is there any downside to using a map instead? I guess the prune operation becomes less efficient, but we could probably keep track of the min block height to make that operation more efficient.

Reply:
There's no practical difference. The number of cached blocks will always be small < 50, so time complexity is no issue. I have no preference between keeping it as a map vs slice